### PR TITLE
fix more tests to work with http2

### PR DESCRIPTION
**Pull request by wfurt**
this is test only change.
This updates tests to use GenericLoopback. 
Removed obsolete special cases for IsNetfxHandler.
Attempt to improve AcceptConnectionAsync. I think this did not work before because of missing .ConfigureAwait(false)

### CODE REVIEW

File: src/Common/tests/System/Net/Http/LoopbackServer.cs [hunk 1]
```diff
@@ -122,34 +122,9 @@ public async Task<Connection> EstablishConnectionAsync()
 
         public async Task AcceptConnectionAsync(Func<Connection, Task> funcAsync)
         {
-            using (Socket s = await _listenSocket.AcceptAsync().ConfigureAwait(false))
+            using (var connection = await EstablishConnectionAsync().ConfigureAwait(false))
```
Comment (davidsh) [inline src/Common/tests/System/Net/Http/LoopbackServer.cs]: `var` should be explicit type (since it isn't obvious).

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 2]
```diff
@@ -79,20 +79,26 @@ public async Task PostAsync_CancelDuringRequestContentSend_TaskCanceledQuickly(b
         [MemberData(nameof(TwoBoolsAndCancellationMode))]
         public async Task GetAsync_CancelDuringResponseHeadersReceived_TaskCanceledQuickly(bool chunkedTransfer, bool connectionClose, CancellationMode mode)
         {
+            if (LoopbackServerFactory.IsHttp2 && (chunkedTransfer || connectionClose))
+            {
+                // There is no chunked encoding in HTTP/2
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: The comment only mentions to skip the test if it is chunked encoding.  But the 'if' statement checks for 'connectionClose'.  How is that related to chunked encoding?  Maybe the comment should be revised?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 3]
```diff
@@ -123,25 +129,33 @@ public async Task GetAsync_CancelDuringResponseHeadersReceived_TaskCanceledQuick
         [MemberData(nameof(TwoBoolsAndCancellationMode))]
         public async Task GetAsync_CancelDuringResponseBodyReceived_Buffered_TaskCanceledQuickly(bool chunkedTransfer, bool connectionClose, CancellationMode mode)
         {
+            if (LoopbackServerFactory.IsHttp2 && (chunkedTransfer || connectionClose))
+            {
+                // There is no chunked encoding or connection header in HTTP/2
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Same question as above.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 4]
```diff
@@ -171,30 +185,38 @@ public async Task GetAsync_CancelDuringResponseBodyReceived_Buffered_TaskCancele
...
         public async Task GetAsync_CancelDuringResponseBodyReceived_Unbuffered_TaskCanceledQuickly(bool chunkedTransfer, bool connectionClose, bool readOrCopyToAsync)
         {
-            if (IsNetfxHandler || IsCurlHandler)
+            if (IsCurlHandler)
             {
                 // doesn't cancel
                 return;
             }
 
+            if (LoopbackServerFactory.IsHttp2 && (chunkedTransfer || connectionClose))
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Same question as above.

File: src/Common/tests/System/Net/Http/LoopbackServer.cs [hunk 5]
```diff
@@ -775,18 +752,26 @@ public override async Task SendResponseAsync(HttpStatusCode statusCode = HttpSta
...
                         }
+                        else if (headerData.Name.Equals("Transfer-Encoding", StringComparison.OrdinalIgnoreCase) && headerData.Value.Equals("chunked", StringComparison.OrdinalIgnoreCase))
+                        {
+                            isChunked = true;
+                        }
                     }
                 }
 
-                if (contentLength < 0)
+                if (!_headersSent)
```
Comment (geoffkizer) [inline src/Common/tests/System/Net/Http/LoopbackServer.cs]: This seems dangerous. What's the need for this?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 6]
```diff
@@ -123,25 +129,33 @@ public async Task GetAsync_CancelDuringResponseHeadersReceived_TaskCanceledQuick
...
 
                     Task serverTask = server.AcceptConnectionAsync(async connection =>
                     {
-                        await connection.ReadRequestHeaderAndSendCustomResponseAsync(
-                            $"HTTP/1.1 200 OK\r\n" +
-                            $"Date: {DateTimeOffset.UtcNow:R}\r\n" +
-                            (!chunkedTransfer ? "Content-Length: 20\r\n" : "") +
-                            (connectionClose ? "Connection: close\r\n" : "") +
-                            $"\r\n123"); // "123" is part of body and could either be chunked size or part of content-length bytes, both incomplete
+                        var headers = new HttpHeaderData[connectionClose ? 2 : 1];
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Why not use a List<> here? Seems simpler.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 7]
```diff
@@ -79,20 +79,26 @@ public async Task PostAsync_CancelDuringRequestContentSend_TaskCanceledQuickly(b
...
                 {
                     var partialResponseHeadersSent = new TaskCompletionSource<bool>();
                     var clientFinished = new TaskCompletionSource<bool>();
 
                     Task serverTask = server.AcceptConnectionAsync(async connection =>
                     {
-                        await connection.ReadRequestHeaderAndSendCustomResponseAsync(
-                            $"HTTP/1.1 200 OK\r\nDate: {DateTimeOffset.UtcNow:R}\r\n"); // missing final \r\n so headers don't complete
+                        await connection.ReadRequestDataAsync();
+                        await connection.SendResponseAsync(HttpStatusCode.OK,isFinal : false);
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Nit: previous test sent a Date header. Do we think that is immaterial?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 8]
```diff
@@ -123,25 +129,33 @@ public async Task GetAsync_CancelDuringResponseHeadersReceived_TaskCanceledQuick
...
-                            $"\r\n123"); // "123" is part of body and could either be chunked size or part of content-length bytes, both incomplete
+                        var headers = new HttpHeaderData[connectionClose ? 2 : 1];
+                        headers[0] = chunkedTransfer ? new HttpHeaderData("Transfer-Encoding", "chunked") : new HttpHeaderData("Content-Length", "20");
+                        if (connectionClose)
+                        {
+                            headers[1] = new HttpHeaderData("Connection", "close");
+                        }
 
+                        await connection.ReadRequestDataAsync();
+                        await connection.SendResponseAsync(HttpStatusCode.OK, headers : headers, body : "123", isFinal : false);
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: no space before the colons for named args

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 9]
```diff
@@ -171,30 +185,38 @@ public async Task GetAsync_CancelDuringResponseBodyReceived_Buffered_TaskCancele
...
 
                     Task serverTask = server.AcceptConnectionAsync(async connection =>
                     {
-                        await connection.ReadRequestHeaderAndSendCustomResponseAsync(
-                            $"HTTP/1.1 200 OK\r\n" +
-                            $"Date: {DateTimeOffset.UtcNow:R}\r\n" +
-                            (!chunkedTransfer ? "Content-Length: 20\r\n" : "") +
-                            (connectionClose ? "Connection: close\r\n" : "") +
-                            $"\r\n");
+                        var headers = new HttpHeaderData[connectionClose ? 2 : 1];
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Same comment as above (use List<>)

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 10]
```diff
@@ -171,30 +185,38 @@ public async Task GetAsync_CancelDuringResponseBodyReceived_Buffered_TaskCancele
...
+                        headers[0] = chunkedTransfer ? new HttpHeaderData("Transfer-Encoding", "chunked") : new HttpHeaderData("Content-Length", "20");
+                        if (connectionClose)
+                        {
+                            headers[1] = new HttpHeaderData("Connection", "close");
+                        }
 
+                        await connection.ReadRequestDataAsync();
+                        await connection.SendResponseAsync(HttpStatusCode.OK, headers : headers, body : "", isFinal : false);
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: colons

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 11]
```diff
@@ -255,15 +272,11 @@ public async Task GetAsync_CancelPendingRequests_DoesntCancelReadAsyncOnResponse
...
                     Task serverTask = server.AcceptConnectionAsync(async connection =>
                     {
-                        await connection.ReadRequestHeaderAndSendCustomResponseAsync(
-                            $"HTTP/1.1 200 OK\r\n" +
-                            $"Date: {DateTimeOffset.UtcNow:R}\r\n" +
-                            $"Content-Length: {contentLength}\r\n" +
-                            $"\r\n");
-
+                                                                                                                                                                                                            await connection.ReadRequestDataAsync();
+                                                                                                                                                                                                            await connection.SendResponseAsync(HttpStatusCode.OK, headers : new HttpHeaderData[] { new HttpHeaderData("Content-Length", contentLength.ToString()) } , body : "", isFinal : false);
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: colons

Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Why is isFinal false here?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 12]
```diff
@@ -255,15 +272,11 @@ public async Task GetAsync_CancelPendingRequests_DoesntCancelReadAsyncOnResponse
...
                         for (int i = 0; i < responseSegments; i++)
                         {
-                            await connection.Writer.WriteAsync(responseContentSegment);
+                            await connection.SendResponseBodyAsync(responseContentSegment, isFinal : i == responseSegments - 1);
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: colon

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 13]
```diff
@@ -255,15 +272,11 @@ public async Task GetAsync_CancelPendingRequests_DoesntCancelReadAsyncOnResponse
 
                     Task serverTask = server.AcceptConnectionAsync(async connection =>
                     {
-                        await connection.ReadRequestHeaderAndSendCustomResponseAsync(
-                            $"HTTP/1.1 200 OK\r\n" +
-                            $"Date: {DateTimeOffset.UtcNow:R}\r\n" +
-                            $"Content-Length: {contentLength}\r\n" +
-                            $"\r\n");
-
+                                                                                                                                                                                                            await connection.ReadRequestDataAsync();
```
Comment (krwq) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: nit: formatting seems to be broken here

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs [hunk 14]
```diff
@@ -171,30 +185,38 @@ public async Task GetAsync_CancelDuringResponseBodyReceived_Buffered_TaskCancele
...
+            if (IsCurlHandler)
             {
                 // doesn't cancel
                 return;
             }
 
+            if (LoopbackServerFactory.IsHttp2 && (chunkedTransfer || connectionClose))
+            {
+                // There is no chunked encoding or connection header in HTTP/2
+                return;
```
Comment (krwq) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: `throw new SkipTestException("Not supported on HTTP/2");`

File: src/Common/tests/System/Net/Http/LoopbackServer.cs (hunk 5 above)
Comment (wfurt) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: there are tests sending incomplete headers and this function can possibly be called multiple times. Only the first call should prepend status to the headers. .

Comment (geoffkizer) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: Where and why is it called multiple times?

It seems bad to store this on the connection. Won't this screw things up for the next request?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs (hunk 7 above)
Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: There is difference between many tests where HTTP/1 does send Date but HTTP/2 does not. If you look at GetAsync_EmptyResponseHeader_Success() change, we special case it for each protocol version. 
I don't think it matters in this test but we should decide and either do not send it unless needed or send it for both protocol versions. 
Rcommendations?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs (hunk 3 above)
Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: This comment is right as Http2 does not really have "Connection: close" and connection management is independent of individual requests. 
It would be nice IMHO if we do not generate incompatible parameter set for HTTP2 but so far I did not find any good way how to do that. If anybody has recommendation, please let me know.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs (hunk 7 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: It's weird that we special case this for GetAsync_EmptyResponseHeader_Success. 

We should be consistent in the implementation of our loopback servers. Either both add Date:, or neither do. I suspect we originally didn't add Date: for HTTP2 because we didn't have very sophisticated header encoding logic. But now we do, so we should be consistent. I don't have a strong opinion on which is best, though.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs (hunk 11 above)
Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: It does not matter for HTTP1. But for Http2 we are trying to avoid sending endStream.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs (hunk 7 above)
Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: ok. In that case I'm leaning towards adding it for both cases to make responses somewhat more realistic. It seems like real servers do use it in both cases.

Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Cancellation.cs]: Seems reasonable to me

**Comment by wfurt** [comment]
feedback from review.
I fixed more tests to get broader prospective. 

I'm not sure about some changes with content-type  changes but alternative would be to always set it explicitly and that would complicate all tests.

COREFX_HTTP2_FORCEUNENCRYPTEDLOOPBACK now works for all tests. e.g. when HttpRequest version is set via helper all, functions can now handle HTTP2 properly.

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs [hunk 15]
```diff
@@ -680,11 +682,68 @@ public override Task<Byte[]> ReadRequestBodyAsync()
...
+                    headers = newHeaders;
+                }
+            }
+
             int streamId = requestId == 0 ? _lastStreamId : requestId;
             bool endHeaders = body != null || isFinal;
-            return SendResponseHeadersAsync(streamId, endStream : isFinal, statusCode, isTrailingHeader : false, endHeaders : endHeaders, headers);
+
+            // TODO: Header continuation support.
+            Assert.False(!_shouldSendStatus && headers != null);
```
Comment (krwq) [inline src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: this would be easier to read as positive assert :smile:

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs [hunk 16]
```diff
@@ -680,11 +682,68 @@ public override Task<Byte[]> ReadRequestBodyAsync()
...
+            if (string.IsNullOrEmpty(body))
+            {
+                await SendResponseHeadersAsync(streamId, endStream: isFinal, statusCode, isTrailingHeader: false, endHeaders: endHeaders, headers);
+                _shouldSendStatus = false;
+            }
+            else
+            {
+                if (headers != null || _shouldSendStatus)
```
Comment (krwq) [inline src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: assuming you got assert just above, does it make sense to check it again?

File: src/Common/tests/System/Net/Http/LoopbackServer.cs [hunk 17]
```diff
@@ -742,40 +743,61 @@ public override async Task SendResponseAsync(HttpStatusCode statusCode = HttpSta
...
 
                 if (headers != null)
                 {
+                    // Process given headers and look for some well-known cases.
                     foreach (HttpHeaderData headerData in headers)
                     {
-                        headerString = headerString + $"{headerData.Name}: {headerData.Value}\r\n";
                         if (headerData.Name.Equals("Content-Length", StringComparison.OrdinalIgnoreCase))
                         {
+                            hasContentLength = true;
```
Comment (krwq) [inline src/Common/tests/System/Net/Http/LoopbackServer.cs]:
```
hasContentLength = headerData.Name.Equals("Content-Length", StringComparison.OrdinalIgnoreCase)
if (hasContentLength)
```

assuming you cannot have more than one content-length - if you cannot it might also be worth to add an assert against more than 1 occurrence. Disregard if it's allowed

Comment (krwq) [reply src/Common/tests/System/Net/Http/LoopbackServer.cs]: would it make sense to actually convert contentLength to `Nullable<int>`?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 18]
```diff
@@ -1978,18 +1993,26 @@ public async Task SendAsync_1xxResponsesWithHeaders_InterimResponsesHeadersIgnor
...
 
-                    var result = new char[TestString.Length];
-                    await connection.ReadBlockAsync(result, 0, TestString.Length);
-                    Assert.Equal(TestString, new string(result));
+                    await connection.SendResponseAsync(responseStatusCode, headers:  new HttpHeaderData[] {
+                        new HttpHeaderData("Cookie", "ignore_cookie=choco2"),
+                        new HttpHeaderData("Content-type", "text/plain"),
+                        new HttpHeaderData("Set-Cookie", SetCookieIgnored2)}, body: "", isFinal: false);
+
+                    //byte[] body = await connection.ReadRequestBodyAsync();
```
Comment (krwq) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: nit: dead code

File: src/Common/tests/System/Net/Http/GenericLoopbackServer.cs [hunk 19]
```diff
@@ -42,6 +42,18 @@ public abstract class GenericLoopbackServer : IDisposable
         public abstract Task AcceptConnectionAsync(Func<GenericLoopbackConnection, Task> funcAsync);
 
         public abstract void Dispose();
+
+        public async Task<HttpRequestData> AcceptConnectionSendResponseAndCloseAsync(HttpStatusCode statusCode = HttpStatusCode.OK, string content = "")
```
Comment (geoffkizer) [inline src/Common/tests/System/Net/Http/GenericLoopbackServer.cs]: How is this different than HandleRequestAsync?

File: src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs [hunk 20]
```diff
@@ -23,6 +23,7 @@ public class Http2LoopbackConnection : GenericLoopbackConnection
         private bool _ignoreWindowUpdates;
         public static TimeSpan Timeout => Http2LoopbackServer.Timeout;
         private int _lastStreamId;
+        private bool _shouldSendStatus;
```
Comment (geoffkizer) [inline src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: As I mentioned previously, it seems like a bad idea to store this kind of state on the connection object itself. It can easily lead to the connection getting into an unexpected state for nonobvious reasons.

If you need to change the behavior of some particular method for some reason, it's better to pass in a flag to that specific function.

Comment (wfurt) [reply src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: I was thinking about extra parameters but this seems to work quite well. For now, we only have tests with one requests/response. Even when we have more, the logic is simple as we need to set this every time we get new request. Any special complicated cases can be written using protocol specific loopback while keeping GenericLoopback relatively simple.

Comment (geoffkizer) [reply src/Common/tests/System/Net/Http/Http2LoopbackConnection.cs]: I still think it's a really bad idea.

Is it really that much more work to change this so we don't need to store state on the connection? If so, can you explain to me why?

File: src/Common/tests/System/Net/Http/GenericLoopbackServer.cs (hunk 19 above)
Comment (wfurt) [reply src/Common/tests/System/Net/Http/GenericLoopbackServer.cs]: I missed the similarity. I don't know why two functions existed in LoopBack server. AcceptConnectionSendResponseAndCloseAsync() seems more descriptive (per earlier @krwq comment) so I left it that calling HandleRequestAsync().

**Comment by wfurt** [comment]
I re-worked the code logic to avoid state variables on connection. All tests with GenericLoopback are still passing. We may do more in next round.

**Comment by wfurt** [comment]
Linux CI failed on System.Text.RegularExpressions

**Comment by eiriktsarpalis** [comment]
@wfurt this PR has stagnated for a while. Could you take a look please?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 21]
```diff
@@ -1440,20 +1454,24 @@ public async Task ReadAsStreamAsync_HandlerProducesWellBehavedResponseStream(boo
...
                 {
-                    await connection.ReadRequestHeaderAsync();
-                    await connection.Writer.WriteAsync("HTTP/1.1 200 OK\r\n");
+                    await connection.ReadRequestDataAsync();
                     switch (chunked)
                     {
                         case true:
-                            await connection.Writer.WriteAsync("Transfer-Encoding: chunked\r\n\r\n3\r\nhel\r\n8\r\nlo world\r\n0\r\n\r\n");
+                            await connection.SendResponseAsync(HttpStatusCode.OK, headers: new HttpHeaderData[] { new HttpHeaderData("Transfer-Encoding", "chunked") }, isFinal: false);
+                            //await connection.SendResponseAsync(body: "3\r\nhel\r\n8\r\nlo world\r\n0\r\n\r\n");
```
Comment (krwq) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: nit: commented out code

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 22]
```diff
@@ -1440,20 +1454,24 @@ public async Task ReadAsStreamAsync_HandlerProducesWellBehavedResponseStream(boo
...
-                            await connection.Writer.WriteAsync("Content-Length: 11\r\n\r\nhello world");
+                            await connection.SendResponseAsync(HttpStatusCode.OK, headers: new HttpHeaderData[] { new HttpHeaderData("Content-Length", "11")}, content: "hello world");
                             break;
 
                         case null:
-                            await connection.Writer.WriteAsync("\r\nhello world");
+                            // This inject Content-Length header with null value to hint Loopback code to not include one automatically.
+                            await connection.SendResponseAsync(HttpStatusCode.OK, headers: new HttpHeaderData[] { new HttpHeaderData("Content-Length", null)}, isFinal: false);
+                            await connection.SendResponseBodyAsync("hello world");
+                            //await connection.SendResponseAsync(body: "hello world");
```
Comment (krwq) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: ditto

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 23]
```diff
@@ -2078,13 +2103,15 @@ public async Task SendAsync_MultipleExpected100Responses_ReceivesCorrectResponse
             {
                 await server.AcceptConnectionAsync(async connection =>
                 {
+                    await connection.ReadRequestDataAsync(readBody: false);
                     // Send multiple 100-Continue responses.
-                    await connection.ReadRequestHeaderAndSendCustomResponseAsync(
-                        string.Concat(Enumerable.Repeat(Valid100ContinueResponse, 3)));
+                    for (int count=0 ; count < 4; count++)
```
Comment (krwq) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: nit: spaces around `=`

**Comment by wfurt** [comment]
all outerloop failures seems unrelated. FileSystemWatcher should be eventually fixed by #40543 
This is ready for final review. All inner loop tests are passing.
